src/main/java/org/springframework/integration/aws/inbound/kinesis/KclMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,408 @@
+		RecordProcessorFactory recordProcessorFactory = new RecordProcessorFactory();
+
+		ConfigsBuilder configsBuilder = new ConfigsBuilder(this.stream, this.consumerGroup,
+				this.kinesisClient, this.dynamoDBClient, this.cloudWatchClient, workerId, recordProcessorFactory);
+		configsBuilder.retrievalConfig().initialPositionInStreamExtended(this.streamInitialSequence);
+		configsBuilder.retrievalConfig().listShardsBackoffTimeInMillis(this.consumerBackoff);
+		configsBuilder.coordinatorConfig().parentShardPollIntervalMillis(this.idleBetweenPolls);
+
+		this.scheduler = new Scheduler(configsBuilder.checkpointConfig(),
+			configsBuilder.coordinatorConfig(),
+			configsBuilder.leaseManagementConfig(),
+			configsBuilder.lifecycleConfig(),
+			configsBuilder.metricsConfig(),
+			configsBuilder.processorConfig(),
+			configsBuilder.retrievalConfig());
+	}
+
+	@Override
+	protected void doStart() {
+		super.doStart();
+		this.executor.execute(() -> this.scheduler.run());
+	}
+
+	/**
+	 * Takes no action by default. Subclasses may override this if they need
+	 * lifecycle-managed behavior.
+	 */
+	@Override
+	protected void doStop() {
+
+		super.doStop();
+		this.scheduler.shutdown();
+
+	}
+
+	private AbstractIntegrationMessageBuilder<Object> prepareMessageForRecord(KinesisClientRecord record) {
+		ByteBuffer data = record.data();
+		byte[] dataArray = new byte[data.remaining()];
+		Object payload = dataArray;
+		data.get(dataArray);
+		Message<?> messageToUse = null;
+
+		if (KclMessageDrivenChannelAdapter.this.embeddedHeadersMapper != null) {
+			try {
+				messageToUse = KclMessageDrivenChannelAdapter.this.embeddedHeadersMapper.toMessage((byte[]) payload);
+
+				payload = messageToUse.getPayload();
+			}
+			catch (Exception e) {
+				logger.warn("Could not parse embedded headers. Remain payload untouched.", e);
+			}
+		}
+
+		AbstractIntegrationMessageBuilder<Object> messageBuilder = getMessageBuilderFactory().withPayload(payload)
+				.setHeader(AwsHeaders.RECEIVED_PARTITION_KEY, record.partitionKey())
+				.setHeader(AwsHeaders.RECEIVED_SEQUENCE_NUMBER, record.sequenceNumber());
+
+		if (messageToUse != null) {
+			messageBuilder.copyHeadersIfAbsent(messageToUse.getHeaders());
+		}
+
+		return messageBuilder;
+	}
+
+	private void performSend(AbstractIntegrationMessageBuilder<?> messageBuilder, Object rawRecord) {
+
+		Message<?> messageToSend = messageBuilder.build();
+		try {
+			sendMessage(messageToSend);
+		}
+		catch (Exception e) {
+			logger.error("Got an exception during sending a '" + messageToSend + "'" + "\nfor the '" + rawRecord
+					+ "'.\n" + "Consider to use 'errorChannel' flow for the compensation logic.", e);
+		}
+	}
+
+	public void setStreamInitialSequence(InitialPositionInStream streamInitialSequence) {
+		setStreamInitialSequence(InitialPositionInStreamExtended.newInitialPosition(streamInitialSequence));
+	}
+
+	public void setStreamInitialSequence(InitialPositionInStreamExtended streamInitialSequence) {
```

Comment:
We can't have two setters with the same name.
Consider to rename this one to an `setStreamInitialSequenceExtended()`.